### PR TITLE
feat(java): support json extraction by scanning

### LINKED ISSUE #5744: [Bug lanceJava Sdk] json_extract fails with InvalidJsonb when querying Utf8 or LargeBinary columns containing raw JSON strings
In the Python SDK, users can easily store JSON as strings and query them using json_extract or json_path expressions. However, in the Java SDK (v2.0.0-beta.9), attempting to use json_extract on a column stored as Utf8 or LargeBinary (containing standard UTF-8 JSON bytes) results in a CDataJniException with an InvalidJsonb error.

This creates a high barrier to entry for Java developers who want to work with semi-structured data without manually defining complex nested Struct schemas.

To Reproduce
Create a dataset with a LargeBinary or Utf8 field in Java.

Insert a valid JSON string as bytes: {"id": 1, "user": {"theme": "dark"}}.

Attempt to scan the dataset with a filter: json_extract(data, '$.user.theme') = '"dark"'

Actual Result:

Plaintext

java.io.IOException: [errno 22] CDataJniException{errno=22, message=External error: Query Execution error: Error applying filter expression to batch: Execution error: Failed to select values from path '$.user.theme': InvalidJsonb}
Expected Behavior
The Java SDK should either:

Automatically handle the coercion of Utf8/LargeBinary to the internal JSON type during query execution (similar to how the Python SDK or DuckDB handles it).

Provide a utility or documentation on how to write data in the expected JSONB format from Java.

Support CAST(column AS JSON) within the filter expression without throwing InvalidJsonb.

Code Snippet
Java

// Schema
Field dataField = new Field("data", new FieldType(true, new ArrowType.LargeBinary(), null), null);
Schema schema = new Schema(Collections.singletonList(dataField));

// Insertion
LargeVarBinaryVector dataVec = (LargeVarBinaryVector) root.getVector("data");
dataVec.setSafe(0, "{\"user\": {\"theme\": \"dark\"}}".getBytes(StandardCharsets.UTF_8));

// Query (Fails)
ScanOptions options = new ScanOptions.Builder()
    .filter("json_extract(data, '$.user.theme') = '\"dark\"'")
    .build();
Environment
Lance Version: 2.0.0-beta.9

Platform: macOS (Apple Silicon) / Linux

JDK: 17 / 21

Arrow Version: 18.1.0

Additional Context
It appears the underlying DataFusion engine expects the LargeBinary column to already be in a specific JSONB binary format rather than raw JSON text. In Python, pa.json_() handles this abstraction, but no equivalent exists in the Java Arrow bindings or Lance Java SDK.

Mention the logs: Including the specific Rust file path from your error message (/rust/lance/src/io/exec/filtered_read.rs:1117:39) is very helpful for the maintainers to locate the exact check that is failing.

```

import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.LargeVarBinaryVector; 
import org.apache.arrow.vector.VectorSchemaRoot;
import org.apache.arrow.vector.ipc.ArrowReader;
import org.apache.arrow.vector.types.pojo.ArrowType;
import org.apache.arrow.vector.types.pojo.Field;
import org.apache.arrow.vector.types.pojo.FieldType;
import org.apache.arrow.vector.types.pojo.Schema;
import org.lance.*;
import org.lance.ipc.LanceScanner;
import org.lance.ipc.ScanOptions;

import java.nio.charset.StandardCharsets;
import java.util.*;

public class LanceNestedJsonSearchFix {

    public static void main(String[] args) throws Exception {
        // 使用新路径以避免旧Schema干扰
        String tableUri = "data/nested_json_lance_fixa";

        try (BufferAllocator allocator = new RootAllocator()) {
            // 1. 修改 Schema: 必须使用 LargeBinary 类型以匹配 json_extract(LargeBinary, Utf8) 签名
            Field dataField = new Field("data",
                    new FieldType(true, new ArrowType.LargeBinary(), null), null);
            Schema schema = new Schema(Collections.singletonList(dataField));

            // 2. 准备 JSON 数据 (字符串形式)
            List<String> jsonData = Arrays.asList(
                    "{\"id\": 1, \"user\": {\"profile\": {\"name\": \"Alice\", \"settings\": {\"theme\": \"dark\", \"notifications\": true}}, \"scores\": [95, 87, 92]}}",
                    "{\"id\": 2, \"user\": {\"profile\": {\"name\": \"Bob\", \"settings\": {\"theme\": \"light\", \"notifications\": false}}, \"scores\": [88, 91, 85]}}"
            );

            // 3. 写入数据
            try (VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator)) {
                // 使用 LargeVarBinaryVector 对应 LargeBinary 类型
                LargeVarBinaryVector dataVec = (LargeVarBinaryVector) root.getVector("data");
                dataVec.allocateNew(jsonData.size()); // 建议显式分配

                for (int i = 0; i < jsonData.size(); i++) {
                    byte[] bytes = jsonData.get(i).getBytes(StandardCharsets.UTF_8);
                    dataVec.setSafe(i, bytes);
                }
                dataVec.setValueCount(jsonData.size()); // 设置向量的实际值数量
                root.setRowCount(jsonData.size());

                try (ArrowReader reader = new MemoryArrowReader(root)) {
                    Dataset.write()
                            .schema(schema)
                            .allocator(allocator)
                            .reader(reader)
                            .uri(tableUri)
                            .mode(WriteParams.WriteMode.OVERWRITE)
                            .execute();
                }
            }

            System.out.println("数据成功写入至: " + tableUri);

            // 4. 打开数据集并查询
            Dataset dataset = Dataset.open(tableUri, allocator);

            // 查询 1: theme = 'dark'
            // 注意 SQL 中的单引号转义
            String darkFilter = "json_extract(data, '$.user.profile.settings.theme') = '\"dark\"'";
            runQuery(dataset, allocator, darkFilter, "Dark Theme Users");

            // 查询 2: scores 数组长度 >= 3
            String scoreFilter = "json_array_length(data, '$.user.scores') >= 3";
            runQuery(dataset, allocator, scoreFilter, "High Scorers (3+ scores)");
        }
    }

    private static void runQuery(Dataset dataset, BufferAllocator allocator, String filter, String label) throws Exception {
        ScanOptions options = new ScanOptions.Builder()
                .filter(filter)
                .build();

        System.out.println("\n--- " + label + " ---");
        try (LanceScanner scanner = LanceScanner.create(dataset, options, allocator);
             ArrowReader reader = scanner.scanBatches()) {
            while (reader.loadNextBatch()) {
                try (VectorSchemaRoot root = reader.getVectorSchemaRoot()) {
                    // 读取时同样使用 LargeVarBinaryVector
                    LargeVarBinaryVector dataRes = (LargeVarBinaryVector) root.getVector("data");
                    for (int i = 0; i < root.getRowCount(); i++) {
                        // 将二进制转回 String 显示
                        System.out.println("Result: " + new String(dataRes.get(i), StandardCharsets.UTF_8));
                    }
                }
            }
        }
    }
}

```

console log
```
LF4J(I): Actual provider is of type [org.slf4j.impl.JBossSlf4jServiceProvider@5f2108b5]
WARNING: A terminally deprecated method in sun.misc.Unsafe has been called
WARNING: sun.misc.Unsafe::allocateMemory has been called by io.netty.util.internal.PlatformDependent0$2 (file:/Users/mullerzhang/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/io/netty/netty-common/4.1.128.Final/netty-common-4.1.128.Final.jar)
WARNING: Please consider reporting this to the maintainers of class io.netty.util.internal.PlatformDependent0$2
WARNING: sun.misc.Unsafe::allocateMemory will be removed in a future release
WARNING: A restricted method in java.lang.System has been called
WARNING: java.lang.System::load has been called by io.questdb.jar.jni.JarJniLoader in an unnamed module (file:/Users/mullerzhang/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/questdb/jar-jni/1.1.1/jar-jni-1.1.1.jar)
WARNING: Use --enable-native-access=ALL-UNNAMED to avoid a warning for callers in this module
WARNING: Restricted methods will be blocked in a future release unless native access is enabled

[2026-01-18T12:22:24Z WARN  lance::dataset::write::insert] No existing dataset at data/nested_json_lance_fixa, it will be created
Exception in thread "main" java.io.IOException: [errno 22] CDataJniException{errno=22, message=External error: Query Execution error: Error applying filter expression to batch: Execution error: Failed to select values from path '$.user.profile.settings.theme': InvalidJsonb, /Users/runner/work/lance/lance/rust/lance/src/io/exec/filtered_read.rs:1117:39}
数据成功写入至: data/nested_json_lance_fixa

--- Dark Theme Users ---
	at org.apache.arrow.c.ArrowArrayStream.getNext(ArrowArrayStream.java:146)
	at org.apache.arrow.c.ArrowArrayStreamReader.loadNextBatch(ArrowArrayStreamReader.java:64)
	at LanceNestedJsonSearchFix.runQuery(LanceNestedJsonSearchFix.java:83)
	at LanceNestedJsonSearchFix.main(LanceNestedJsonSearchFix.java:67)


```

### PR DESCRIPTION
**Pull request by majin1102**
Close #5744

For now arrow-java does't provide a convenient way like `pa.json()`. So we extend the field type in lance java binding to support related json udf

### CODE REVIEW

**Comment by majin1102** [comment]
@Xuanwo Could you take a look at this json related feat for Java binding? 

Thanks

File: java/src/test/java/org/lance/JsonExtractionTest.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import java.nio.file.Path;
+import java.util.Arrays;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class JsonExtractionTest {
+
+  @Test
+  void testJsonExtraction(@TempDir Path tempDir) throws Exception {
```
Comment (yanghua) [inline java/src/test/java/org/lance/JsonExtractionTest.java:48]: Can we add more tests for some corner cases? For example, `null` field, invalid JSON string, `jsonLargeUtf8 `.

**Comment by majin1102** [comment]
Ready for another look @yanghua
